src/flowcean/environments/ode_environment.py

```diff
@@ -1,109 +1,161 @@
+        """Create a state from a numpy array.
 
-class ODEEnvironment(IncrementalEnvironment):
-    r"""Environment to get samples by integrate a differential equation.
+        Args:
+            state: State as a numpy array.
 
-    This environment generates samples by integrating a differential equation
-    of the form
+        Returns:
+            State instance.
+        """
+
+
+class OdeSystem[X: State](ABC):
+    r"""System governed by a ordinary differential equation.
+
+    This class represents a continuous system. The system is defined by a
+    differential flow function $f$ that governs the evolution of the state $x$.
 
     $$
     \begin{aligned}
-        \dot{\vec{x}} &= f(t, \vec{x}) \\
-        \vec{y} &= g(\vec{x})
+        \dot{x} &= f(t, x) \\
+        y &= g(x)
     \end{aligned}
     $$
 
-    where $f$ is the differential function, $\vec{x}$ the current state,
-    $\dot{\vec{x}}$ it's derivative and $\vec{y}$ is the output calculated by
-    the output function $g$.
-
-    Args:
-        f: Handle to the ODE function.
-        x0: Initial state for the integration.
-        g: Output function. Defaults to an invariant function.
-        tstep: Time between two samples yielded by `get_next_data`.
-        output_names: Names of the output features. Defaults to `y_0`, `y_1`
-            ... `y_n`
-        implicit_integration: Indicate if an implicit integrator should be
-            used. If `True` the `BDF` integrator will be used. Otherwise `RK45`
-            is used for integration.
-
-    Example:
-        Create an environment that integrates a simple pendulum with length
-        $L=1$. Samples are taken every 10 ms with output columns named `phi`
-        and `Dphi`.
-        ```python
-        environment = ODEEnvironment(
-            lambda _, x, L=1, g=9.81: np.array(
-                [x[1], -g/L * np.sin(x[0])]
-            ),
-            np.array([np.sin(np.deg2rad(30)), 0]),
-            tstep=1e-2,
-            output_names=["phi", "Dphi"],
-        )
-        ```
+    The system can be integrated to obtain the state at a future time.
+
+    Attributes:
+        t: Current time.
+        state: Current state.
     """
 
+    t: float
+    state: X
+
+    def __init__(
+        self,
+        t: float,
+        state: X,
+    ) -> None:
+        """Initialize the system.
+
+        Args:
+            t: Initial time.
+            state: Initial state.
+        """
+        self.t = t
+        self.state = state
+
+    @abstractmethod
+    def flow(
+        self,
+        t: float,
+        state: NDArray[np.float64],
+    ) -> NDArray[np.float64]:
+        """Ordinary differential equation.
+
+        Compute the derivative of the state $x$ at time $t$.
+
+        Args:
+            t: Time.
+            state: State.
+
+        Returns:
+            Derivative of the state.
+        """
+
+    def step(
+        self,
+        dt: float,
+    ) -> tuple[Sequence[float], Sequence[X]]:
```

Comment:
The problem with that is: The step function is advancing the system by a duration of `dt`. But the integration might have done multiple steps to actually reach that time. This function thus returns the entire history that was evaluated. Otherwise this information is lost. Especially for non-linear systems this is very necessary.